
# Convert dds from DE

Let's check the number of assigned reads per sample as a proxy of the sequencing depth in the main dds object.

```{r dds-from-gse}
if (!purrr::pluck(params, "load_dds_objects", .default = FALSE) && 
    purrr::pluck(params, "load_gse", .default = TRUE)) {
  (dds <- DESeq2::DESeqDataSet(gse, design = design_definition))
 } else {
  warning("No dds object created from gse, this may cause problems in the workflow, ignore if it was done with intention")
}
```

# Gene annotation

To create the gene annotation tables, only if first run. (if create_annotation_df = TRUE)

```{r gene_annotation, eval = FALSE}
if (purrr::pluck(params, "create_annotation_df", .default = FALSE)) create_annotations(dds)
```