---
title: "Introductory analysis"
author: "Edoardo Filippi"
execute:
  echo: true
  warning: false
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    number_sections: true
    theme: cosmo

---
How this works: 

With these Parameters you can control the analysis

The workflow is supposed to be: run it once setting the filtering that you want and the contrasts that you need, the objects will be saved in the "experiment_name" sub-folder in the analysis results folder. When you want to run it again with the same parameters and, for example, create a report, you can set the load_dds_objects to true, they will be loaded and unnecessary operations skipped.

```{r params, eval = TRUE}
params <- list(
  "filter_complete_dds" = FALSE,
  "load_complete_dds" = FALSE,
  "load_dds_objects" = TRUE,
  "save_dds_objects" = FALSE,
  "complete_dds_source" = "Analyses_data/Complete_dds/dds_at2_cleaned.RDS",
  "experiment_name" = "explorative_data_analysis",
  "experiment_metadata_file" = "samplesDesign_paul_AT2full.xlsx",
  "create_annotation_df" = FALSE
)
```

# setup

This is the function to modify when you want to filter the object in a specific way before beginning. 
All the subsequent analyses will be carried out starting from the samples present in this object.

```{r filtering-object, eval = TRUE, include = TRUE}
filter_dds <- function(dds_to_filter) {

  dds_to_filter <- dds_to_filter[, colData(dds_to_filter)$project_scenario == "kidney_ifn_1"]

  dds_to_filter <- dds_to_filter[, colData(dds_to_filter)$group %in% c(group_of_interest,"wt")]

  return(dds_to_filter)
}
```


```{r setup, include=FALSE, cache=FALSE, eval = TRUE, echo = FALSE}
library(knitr)
opts_chunk$set(
  fig.path = paste0("./figures/", purrr::pluck(params, "experiment_name", .default = FALSE)),
  fig.align = 'center',
  fig.show = 'asis',
  eval = TRUE,
  fig.width = 10,
  fig.height = 7,
  tidy = FALSE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  size = 'small',
  comment = "##",
  echo = TRUE,
  results = 'markup',
  dev = c("png", "pdf")
)
options(replace.assign = TRUE,
        width = 80)
```

```{r loadLibraries,results='hide'}
library("SummarizedExperiment")
library("ggplot2")
# library("limma")
library("pcaExplorer")
library("DESeq2")
library("GeneTonic")
library("DT")
library("org.Mm.eg.db")
library("biomaRt")
# library("clusterProfiler")
library("topGO")
# library("pcaExplorer")
```

```{r load_functions}
source("load_helper_functions.r")
```

# Importing files into R and preparing workspace

We start by importing all the necessary data.
These variables will be created in the workspace:

* anno_df
* anns
* complete_dds (if needed)
* experiment_metadata

If anno_df and anns still need to be created they will not be loaded

```{r load-function}
prepare_workspace(params)
```

If loading parameter set to true load the already analyzed data necessary objects

```{r load-data, eval = TRUE}
if (purrr::pluck(params, "load_dds_objects", .default = FALSE)) 
  load_analyses(params)
```

Filtering using the funciton defined precently 

```{r filter-complete_dds, eval = TRUE}
if (
  purrr::pluck(params, "filter_complete_dds", .default = FALSE) && 
  !purrr::pluck(params, "load_dds_objects", .default = FALSE)
  ) dds <- filter_dds(complete_dds) else if (!purrr::pluck(params, "load_dds_objects", .default = FALSE)) dds <- complete_dds # (if needed)
```

# Gene annotation

To create the gene annotation tables, only if first run. (se create_annotation_df = TRUE)
```{r gene_annotation, eval = FALSE}
if (purrr::pluck(params, "create_annotation_df", .default = FALSE)) create_annotations(dds)
```


# EDA - Exploratory Data Analysis

Let's take a look at the experiment details:

```{r prepare-dds, echo = FALSE}
factor_columns <- c("group")
for (col in factor_columns) {
  colData(dds)[[col]] <- droplevels(factor(colData(dds)[[col]])) 
}
colnames(colData(dds))
for (col in factor_columns) {
  message(paste0("Unique values for ", col, ": ", paste0(unique(colData(dds)[[col]]), collapse = ", ")))
}
message("are all necessary columns factors?")
sapply(colData(dds), is.factor)
```


Are the columns factors? 

```{r ngenes-genes, echo = FALSE}
message(paste0("numer of genes: ", nrow(dds))) # total number of genes
```

Does this gene entry have expression different from 0 in at least one sample?
```{r unexp-genes, echo = FALSE}

table(rowSums(assay(dds, "counts")) != 0)
```
# Defining design

These are the groups:

* M01 -> ctrl
* M02 -> AT2
* M03 -> AT2finerenon
* M04 -> AT2sglt2i
* M05 -> finerenon
* M06 -> sglt2i

We would then like to see these contrasts:

Finerenone part:

* group_AT2_vs_ctrl
* group_AT2finerenon_vs_AT2
* group_AT2finerenon_vs_ctrl
* group_finerenon_vs_ctrl
* group_sglt2i_vs_AT2

SGLT2 part:

* group_AT2_vs_ctrl
* group_AT2sglt2i_vs_AT2
* group_AT2sglt2i_vs_ctrl
* group_sglt2i_vs_ctrl
* group_sglt2i_vs_AT2

```{r defining-contrasts, echo = FALSE}

finerenone_contrasts <- c(
  "group_AT2_vs_ctrl",
  "group_AT2finerenon_vs_AT2",
  "group_AT2finerenon_vs_ctrl",
  "group_finerenon_vs_ctrl",
  "group_sglt2i_vs_AT2"
)

sglt2_contrasts <- c(
  "group_AT2_vs_ctrl",
  "group_AT2sglt2i_vs_AT2",
  "group_AT2sglt2i_vs_ctrl",
  "group_sglt2i_vs_ctrl",
  "group_sglt2i_vs_AT2"
)
```

This means that this design matrix is sufficient to us:

Our design matrix:

```{r experiment-matrix}
# visualizedesign matrix
design_definition <- ~ group

model_matrix <- model.matrix(design_definition, data =  colData(dds))

kable(model_matrix)
```

```{r assigning-design}
design(dds) <- design_definition 
```


# Number of reads

Let's check the number of assigned reads per sample as a proxy of the sequencing depth in the main dds object.

```{r plot-seq-depth}
myd <- data.frame(
  counts = colSums(counts(dds)),
  sample = colnames(dds))
ggplot(myd, aes(x = sample,weight = counts, fill = sample)) +
  geom_bar() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  theme_bw() + 
  coord_flip()
```

# PCA plots

We can observe the clustering of the mouses which show a similar expression profile in the following pca plot.You can change the the value in the intgroup value to modify the column of the experiment_metadata according how you want to group the samples if you wish.

```{r vst, include = FALSE}
vst <- vst(dds)
```

```{r pca}

message(paste0("possible grouping columns: ", paste(factor_columns, collapse = ", ")))

pca_data <- pcaExplorer::pcaplot(vst, returnData = T, intgroup = "group", ellipse = FALSE, text_labels = FALSE, ntop = 5000)

pcaExplorer::pcaplot(vst, intgroup = "group", ellipse = FALSE, text_labels = TRUE, ntop = 5000)
```

PCA explorer can be used to interactively observe the object, you can also delete single objects and explore how the PCA changes. In the source code at this point there is an example of the chunk that you need to run to start the app. 

```{r pcaExplorer, eval = FALSE, echo = FALSE}
pcaExplorer(
  dds = dds,
  annotation = anns,
)
```


# DE - Differential Expression

## DE Analysis

We need to prepare the dds objects in order to be sure we obtain the wished contrasts, and then perform the differential expression analysis with the DESeq2 package. We can see which object can be used to obtain which contrasts.

```{r de-analysis, echo = FALSE}
# if not load dds objects 
if (!purrr::pluck(params, "load_dds_objects", .default = FALSE)){

  # releveling
  dds$group <- relevel(dds$group, "ctrl")

  # de analysis
  dds <- DESeq(dds)

  # relevel
  # dds$group <- relevel(dds$group, ref = "lupus_ATII")
  dds_releveled <- dds
  dds_releveled$group <- relevel(dds_releveled$group, "AT2")
  dds_releveled <- nbinomWaldTest(dds_releveled)
}
```

```{r contrasts}
# contrasts
(contrasts <- resultsNames(dds))
```
```{r contrasts-releveled}
# contrasts
(contrasts_releveled <- resultsNames(dds_releveled))
```

## DE Parameters

```{r FDR}
# set FDR 
FDR = 0.1 # false discovery rate
```

```{r myresuSet,  include = FALSE}
if (!exists("myresuSet")) myresuSet <- list()
```

## Contrasts

You can select for which contrasts to see the DE results, to do this you can copy the following blocks as many times as you want and add the correct contrast name. The correct object also needs to be chosen.

## finerenone_contrasts

<!-- <details>

<summary>click here to display DE results of </summary> -->

```{r, echo = FALSE,include = FALSE}
# You need this code to conduct the magic dependences attaching...
DT::datatable(matrix())
```


```{r, results = 'asis', echo = FALSE}

for (contrast in finerenone_contrasts) {
  if (grepl("AT2$", contrast)) dds_to_use <- dds_releveled else dds_to_use <- dds
  if (!purrr::pluck(params, "load_dds_objects", .default = FALSE))
    myresuSet <- alltheresults(
      myresuSet, dds_to_use,
      contrast = c(contrast), FDR = FDR,
      anno_df = anno_df, anns = anns, species = "Mus_musculus")
  # myresuSet[[contrast]]$maplot_res
    cat(knitr::knit_print(DT::datatable(myresuSet[[contrast]]$etbl_res_DE, escape = FALSE, rownames = FALSE, 
        caption = htmltools::tags$caption(style = 'caption-side: top; color:black;; font-size: 2em', contrast))))
}
rm(dds_to_use)
```



## sglt2_contrasts

```{r, results = 'asis', echo = FALSE}

for (contrast in sglt2_contrasts) {
  if (grepl("AT2$", contrast)) dds_to_use <- dds_releveled else dds_to_use <- dds
  if (!purrr::pluck(params, "load_dds_objects", .default = FALSE))
    myresuSet <- alltheresults(
      myresuSet, dds_to_use,
      contrast = c(contrast), FDR = FDR,
      anno_df = anno_df, anns = anns, species = "Mus_musculus")
  # myresuSet[[contrast]]$maplot_res
    cat(knitr::knit_print(DT::datatable(myresuSet[[contrast]]$etbl_res_DE, escape = FALSE, rownames = FALSE, 
        caption = htmltools::tags$caption(style = 'caption-side: top; color:black; font-size: 2em', contrast))))
}
rm(dds_to_use)
```

# Enrichment
<!-- 
</details>

<details>
## Analysis with topGO -->
```{r setup-enrichment}
expressedInAssay <- (rowSums(assay(dds))>0)
geneUniverseExprENS <- rownames(dds)[expressedInAssay]
geneUniverseExpr <- anno_df$gene_name[match(geneUniverseExprENS, anno_df$gene_id)]
```

```{r enrichment, eval = TRUE, echo = FALSE}
# enrichment with topGO
if (!purrr::pluck(params, "load_dds_objects", .default = FALSE)) {
  for (i in names(myresuSet)) {
    message(i)
    if (nrow(myresuSet[[i]][["tbl_res_DE"]]) > 0) {
      myresuSet[[i]][["topGO_tbl"]] <- 
        topGOtable(DEgenes = myresuSet[[i]][["tbl_res_DE"]]$geneSymbol, 
                  BGgenes = geneUniverseExpr, 
                  ontology = "BP",
                  geneID = "symbol",
                  addGeneToTerms = TRUE, 
                  topTablerows = 500,
                  mapping = "org.Mm.eg.db")
    }
  }
}
```

## finerenone_contrasts
```{r, echo = FALSE,include = FALSE}
# You need this code to conduct the magic dependences attaching...
DT::datatable(matrix())
```

```{r enrichment-tables-1, eval = TRUE, echo = FALSE, results = 'asis'}
for (contrast in finerenone_contrasts){
  cat(
    knitr::knit_print(
      DT::datatable(
        myresuSet[[contrast]]$topGO_tbl, 
        escape = FALSE, 
        rownames = TRUE, 
        caption = htmltools::tags$caption(style = 'caption-side: top; color:black; font-size: 2em', paste0("Enrichment ", contrast)
        )
      )
    )
  )
}
```

## sglt2_contrasts

```{r enrichment-tables-2, eval = TRUE, echo = FALSE, results = 'asis'}
for (contrast in sglt2_contrasts){
  cat(
    knitr::knit_print(
      DT::datatable(
        myresuSet[[contrast]]$topGO_tbl, 
        escape = FALSE, 
        rownames = FALSE, 
        caption = htmltools::tags$caption(style = 'caption-side: top; color:black; font-size: 2em', paste0("Enrichment", contrast)
        )
      )
    )
  )
}
```

# Save 

```{r save, eval=TRUE}
if (purrr::pluck(params, "save_dds_objects", .default = FALSE))
  save_analysis_objects(
      params, 
      myresuSet, 
      dds,
      dds_releveled = dds_releveled, 
      # dds_interaction_term = dds_interaction_term
      )
```

# Gene plots

Here you can see the expression of some genes of your choice. 

Tπhere is a section to select the genes that are present from a specific gene set. You need to select the contrast and the GO term that you wish and the genes will be plotted. you can copy and paste the code if you wish to do this plot for more gene sets

## plot list title
œ
```{r genePlots, eval = TRUE, echo = TRUE}
dds <- add_symbols(dds, anno_df)
contrast <- "group_AT2_vs_ctrl" #or the one you prefer
go_term <- "GO:0071347" # to choose
res_enrichment <- myresuSet$group_AT2_vs_ctrl$topGO_tbl
genes_string <- res_enrichment |> dplyr::filter(GO.ID == go_term) |> dplyr::pull(genes)
genes <- strsplit(genes_string, ",")[[1]]

for (gene in genes) {

  if (gene %in% rowData(dds)$SYMBOL)
    gene_ensemblID <- as.data.frame(rowData(dds)) |> dplyr::filter(SYMBOL == gene) |> rownames()
  else {
    message("Gene not found in the dataset.")
    next
  }

  if (sum(counts(dds)[gene_ensemblID,]) > 0){
    # print(paste0("the following plot is for gene: ", gene))
    print(mosdef::gene_plot(dds, gene_ensemblID, intgroup = "group") + ggtitle(gene))
  }
}

```

```{r startGeneTonic, eval = FALSE, echo = FALSE}
# It actually does not work without the results of the enrichment
gtl <- resuset_to_gtl(myresuSet, "group_IFNaRko_IL28Rko_vs_wt", dds, anno_df, enrich_algo = "topGO_tbl") 
GeneTonic(gtl = gtl)

```

# Session information {}

```{r sessioninfo}
BIUMmisc::sessioninfo_extra()
```
